### PR TITLE
feat(just-cartesian-product): enable esm

### LINKED ISSUE #192: Use ES Modules with type:module
It'd be great to have `just` be a Node-compatible ES Module with `type: "module"` in the package.json

I don't remember if there's a backwards-compatible way to have both CJS and `type:module` in the same import, but this is definitely possible:

```
- index.js // cjs file
- esm/index.js // esm file
- esm/package.json // Content: `{"type": "module"}`
```

This would enable both:

```js
const something = require('just-something');
```

```js
import something from 'just-something/esm'; // Notice the `/esm`
```

### PR DESCRIPTION
**Pull request by raulfdm**
## Related
- closes #192

## General

Enables ESM for all "just" packages.

The PR itself looks huge but that's because I have to generate a `index.esm.js` and tweak package.json from every single package.

## Description
More and more we're migrating to ESM environment and it's important to, at least in this transition moment, support both.

This PR adds Rollup and does a transpiles the original entry point (written in CJS) into a new `index.esm.js` file.

The approach chosen was suggested here and as reference, used the suggestion from Dr. Axel Rauschmayer post: [Hybrid packages (ESM and CommonJS)](https://2ality.com/2019/10/hybrid-npm-packages.html#option-1-(experimental%2C-needs-conditional-exports)%3A-esm-and-commonjs-are-both-bare-imports).

The first attempt was using ESBuild, but after looking the outcome code and a suggestion ([here](https://github.com/angus-c/just/pull/315#issuecomment-950175754), I've moved to Rollup + commonjs plugin.

## Test

To see if works as expected, I've publish the package `just-cartesian-product` into a local npm proxy using [Verdacio](https://verdaccio.org/) (more details of how to do that [here](https://raulmelo.dev/blog/test-publishing-js-package-locally)). Then, I've published that package (with both CJS and ESM) to there:

Screenshot 2021-10-23 at 17 38 00

After that, I've created a brand-new vanilla-ts project using [Vite](https://vitejs.dev/guide/#scaffolding-your-first-vite-project) which is well-known to push ESM ecosystem, install the `just-cartesian-product` from my local registry, copied the code from the README and console it:

Screenshot 2021-10-23 at 17 41 21

Screenshot 2021-10-23 at 17 41 06

After that, I run `yarn build` to see if everything would be picked correctly and run a served the outcome with a web server (live-server):

Screenshot 2021-10-23 at 17 43 32

Screenshot 2021-10-23 at 17 43 26

Worked like a charm! 🕺

## Overall Publishing

After I add `index.esm.js`,add info on each package's json, and commit, I've ran:

```bash
yarn lerna publish  --yes major --registry=http://0.0.0.0:4873/
```

This automatically bumped all versions to major and publish to my local registry:

Screenshot 2021-10-24 at 07 46 59

## Random GIF

Dab!

## Misc.

<details>
<summary>Script used to update package.json's</summary>

```js
const glob = require('glob')
const fs = require('fs')
const path = require('path')

glob('packages/**/package.json', (er, filePaths) => {
  for (let filePath of filePaths) {
    const result = {};
    const {
      name,
      version, description,
      main,
      types,
      scripts,
      ...rest
    } = require(path.resolve(__dirname, filePath))


    result.name = name
    result.version = version;
    result.description = description;
    result.type = 'commonjs',
      result.main = main;
    result.module = 'index.esm.js',
      result.exports = {
        ".": {
          "require": "./index.js",
          "default": "./index.esm.js"
        }
      }

    if (types) {
      result.types = types
    }
    result.scripts = { ...scripts, build: 'rollup -c' }

    const finalContent = { ...result, ...rest }

    fs.writeFileSync(filePath, JSON.stringify(finalContent, null, 2))
  }
});
```
</details>

### CODE REVIEW

**Comment by angus-c** [comment]
Hi Raul, this looks great. I'm excited. Tests are passing now. Can you apply [this diff ](https://github.com/angus-c/just/files/7447653/rauldiff.txt) to your forked repo to resolve conflicts?

Then I think we're good to go!

**Comment by raulfdm** [comment]
@angus-c I've reverted the version bump and manage the conflicts.

Since I can't merge this PR, whenever you feel it's ok, go ahead. I'd recommend you to instead `merge`, `square and merge` you can you have a single commit message and add "BREAKING CHANGE" so people can be aware that's a major (and why).

Also you'll need to run lerna version by yourself.

Anyhow, thanks mate! :)

**Comment by angus-c** [comment]
Thank you! 

(I still think it's not breaking change :) Instead it's minor version bump since acting users can upgrade without anything changing, but it adds a new feature (esm flavor))

**Comment by PuruVJ** [comment]
@angus-c it'd be great to publish all these to npm. Some of em were last published 10 months ago but updated very recently 🥲🥲

**Comment by angus-c** [comment]
@PuruVJ curious which those were, I try to publish everything as soon as it's updated. Anyway I published every package just now :)

**Comment by angus-c** [comment]
@raulfdm ok all landed and published. Seems to work perfectly. 
Will tweet about it tomorrow!

File: config/createRollupConfig.js [hunk 1]
```diff
@@ -0,0 +1,17 @@
...
+
+/**
+ *
+ * @param {string} path
+ */
+module.exports = function createRollupConfig(path) {
+  return {
+    input: resolve(path, './index.js'),
+    output: {
+      file: resolve(path, './index.esm.js'),
```
Comment (JakeChampion) [inline config/createRollupConfig.js:12]: This extension does not work for NodeJS' native esm implementation, the extension needs to be `.mjs` for NodeJS to run the file in ESM mode.


Right now when using NodeJS' native esm implementation, an error will be thrown when using `import` on any Just packages:
```
(node:55016) Warning: To load an ES module, set "type": "module" in the package.json or use the .mjs extension.
/tmp/just-esm-example/node_modules/just-camel-case/index.esm.js:40
export {stringCamelCase as default};
^^^^^^

SyntaxError: Unexpected token 'export'
    at Object.compileFunction (node:vm:352:18)
    at wrapSafe (node:internal/modules/cjs/loader:1031:15)
    at Module._compile (node:internal/modules/cjs/loader:1065:27)
    at Object.Module._extensions..js (node:internal/modules/cjs/loader:1153:10)
    at Module.load (node:internal/modules/cjs/loader:981:32)
    at Function.Module._load (node:internal/modules/cjs/loader:822:12)
    at ModuleWrap.<anonymous> (node:internal/modules/esm/translators:196:29)
    at ModuleJob.run (node:internal/modules/esm/module_job:183:25)
    at async Loader.import (node:internal/modules/esm/loader:178:24)
```

**Pull request merged**